libshaderc_spvc/src/spvc.cc

```diff
@@ -164,74 +174,370 @@ void shaderc_spvc_compiler_release(shaderc_spvc_compiler_t compiler) {
+           std::vector<uint32_t>& out) {
+  if (index < inst->num_operands) {
+    const spv_parsed_operand_t& op = inst->operands[index];
+    assert(op.type == SPV_OPERAND_TYPE_ID);
+    const uint32_t* p = inst->words + op.offset;
+    std::copy(p, p + inst->num_operands - index, std::back_inserter(out));
   }
-  return result;
 }
+
+// Recursive template base case.
+void GetOpsRecur(const spv_parsed_instruction_t* inst, unsigned index) {}
+
+// Recursive template to do any number of GetOp().
+template <typename First, typename... Rest>
+void GetOpsRecur(const spv_parsed_instruction_t* inst, unsigned index,
+                 First& first, Rest&... rest) {
+  GetOp(inst, index, first);
+  GetOpsRecur(inst, index + 1, rest...);
+}
+
+// Start the recursive template.
+template <typename... Outs>
+void GetOps(const spv_parsed_instruction_t* inst, Outs&... outs) {
+  GetOpsRecur(inst, 0, outs...);
+}
+
+// This struct holds a SPIRV-Cross compiler object of type 'CompilerType,'
+// all the state needed while building the spirv_cross::ParsedIR object consumed
+// by the compiler, and a shaderc_spvc_compilation_result_t which indicates
+// the result of validation/parsing/compiling.
+template <class CompilerType>
+struct Compiler {
+  CompilerType* cross = nullptr;
+  spirv_cross::SPIRFunction* current_function = nullptr;
+  spirv_cross::SPIRBlock* current_block = nullptr;
+  spirv_cross::ParsedIR ir;
+  shaderc_spvc_compilation_result_t result =
+      nullptr;  // NOT freed by our destructor
+
+  void consume_validation_message(spv_message_level_t level, const char* src,
+                                  const spv_position_t& pos,
+                                  const char* message) {
+    assert(result);
+    result->messages.append(message);
+    result->messages.append("\n");
+  }
+
+  static spv_result_t parse_header(void* user_data, spv_endianness_t endian,
+                                   uint32_t magic, uint32_t version,
+                                   uint32_t generator, uint32_t id_bound,
+                                   uint32_t reserved) {
+    return ((Compiler*)user_data)
+        ->parse_header(endian, magic, version, generator, id_bound, reserved);
+  }
+
+  spv_result_t parse_header(spv_endianness_t endian, uint32_t magic,
+                            uint32_t version, uint32_t generator,
+                            uint32_t id_bound, uint32_t reserved) {
+    ir.set_id_bounds(id_bound);
+    return SPV_SUCCESS;
+  }
+
+  static spv_result_t parse_instruction(void* user_data,
+                                        const spv_parsed_instruction_t* inst) {
+    return ((Compiler*)user_data)->parse_instruction(inst);
+  }
+
+  // Duplicate behavior of SPIRV-Cross.
+  spv_result_t parse_instruction(const spv_parsed_instruction_t* inst) {
+    switch (inst->opcode) {
+      case spv::OpExtInstImport: {
+        uint32_t id;
+        std::string ext;
+        GetOps(inst, id, ext);
```

Comment:
I don't understand how this template stuff works. Why do we not just end up scribbling all over memory for the string case?

Would this be simpler to just have GetOp<uint32>, GetOp<spv::ExecutionModel> and GetStringOp methods that that you call on the individual lines above instead of using the GetOps template?

Reply:
I would rather write this

 GetOps(inst, op1, op2, op3);

instead of this

 GetOp(inst, op1, 1);
 GetOp(inst, op2, 2);
 GetOp(inst, op3, 3);

or even worse

 GetOpBlah(inst, op1, 1);
 GetOpFoo(inst, op2, 2);
 GetOpExecutionModel(inst, op3, 3);

A pattern that will appear 100 times once all the instruction cases are filled in.

GetOp is specialized to extract characters from the spirv stream and put them in the given std::string.

Reply:
You're already doing:

```
uint32_t id;
std::string ext;
GetOps(inst, id, ext);
```
This would just mean:
```
uint32_t id = GetOp<uint32_t>(0);
std::string ext = GetString(1);
````
(or GetOp<const char*> if that's easier)

Reply:
Still seems like avoidable tedium, writing in types and indices, hundreds of times, just to get rid of one tricky template.  Can I comment it better?
Care to weigh in @dneto0 ?